spring-cloud-gcp-autoconfigure/src/main/java/com/google/cloud/spring/autoconfigure/sql/R2dbcCloudSqlEnvironmentPostProcessor.java

```diff
@@ -0,0 +1,95 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.spring.autoconfigure.sql;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+
+import org.springframework.boot.SpringApplication;
+import org.springframework.boot.env.EnvironmentPostProcessor;
+import org.springframework.core.env.ConfigurableEnvironment;
+import org.springframework.core.env.MapPropertySource;
+import org.springframework.util.Assert;
+import org.springframework.util.ClassUtils;
+
+/**
+ * Builds connection string for Cloud SQL through Spring R2DBC by requiring only a
+ * database and instance connection name.
+ *
+ */
+public class R2dbcCloudSqlEnvironmentPostProcessor implements EnvironmentPostProcessor {
+	private static final Log LOGGER = LogFactory.getLog(R2dbcCloudSqlEnvironmentPostProcessor.class);
+
+	@Override
+	public void postProcessEnvironment(ConfigurableEnvironment environment, SpringApplication application) {
+		if (environment.getPropertySources().contains("bootstrap")) {
+			// Do not run in the bootstrap phase as the user configuration is not available yet
+			return;
+		}
+
+		DatabaseType databaseType = getEnabledDatabaseType(environment);
+		if (databaseType != null) {
+			PropertiesRetriever propertiesRetriever = new PropertiesRetriever(environment);
+			String r2dbcUrl = createUrl(databaseType, propertiesRetriever.getCloudSqlProperties());
+			if (LOGGER.isInfoEnabled()) {
+				LOGGER.info("Default " + databaseType.name()
+						+ " R2dbcUrl provider. Connecting to "
+						+ r2dbcUrl);
+			}
+
+			// Add default username as fallback when not specified
+			Map<String, Object> fallbackMap = new HashMap<>();
+			fallbackMap.put("spring.r2dbc.username", databaseType.getDefaultUsername());
+			environment.getPropertySources().addLast(new MapPropertySource("CLOUD_SQL_R2DBC_USERNAME", fallbackMap));
+
+			Map<String, Object> primaryMap = new HashMap<>();
+			primaryMap.put("spring.r2dbc.url", r2dbcUrl);
+			environment.getPropertySources().addFirst(new MapPropertySource("CLOUD_SQL_R2DBC_URL", primaryMap));
+		}
+	}
+
+	String createUrl(DatabaseType databaseType, GcpCloudSqlProperties sqlProperties) {
+		Assert.hasText(sqlProperties.getDatabaseName(), "A database name must be provided.");
+		Assert.hasText(sqlProperties.getInstanceConnectionName(),
+				"An instance connection name must be provided in the format <PROJECT_ID>:<REGION>:<INSTANCE_ID>.");
+
+		return String.format(databaseType.getR2dbcUrlTemplate(),
+				sqlProperties.getInstanceConnectionName(), sqlProperties.getDatabaseName());
+	}
+
+	DatabaseType getEnabledDatabaseType(ConfigurableEnvironment environment) {
+		if (Boolean.parseBoolean(environment.getProperty("spring.cloud.gcp.sql.enabled", "true"))
```

Comment:
Done. This method now returns `null` if the dependency that is used to enable Spring R2DBC auto-configuration is not present.

Reply:
What if the user has the R2DBC dependency, but doesn't want our auto-configuration to kick in? I think we might need something like `spring.cloud.gcp.sql.r2dbc.enabled`.

Reply:
That would be instead of `spring.cloud.gcp.sql.enabled` for Cloud SQL R2DBC autoconfig being on, not in addition to, right?
I guess theoretically it's possible that an app wants `spring.cloud.gcp.sql.enabled` for blocking connection to Cloud SQL, and then an R2DBC driver to another type of database. It would be weird for a reactive app to have blocking components, but not impossible.

Reply:
Ideally, `spring.cloud.gcp.sql.enabled=false` would disable all of our Cloud SQL starters.
Maybe we need to add two properties: `spring.cloud.gcp.sql.r2dbc.enabled` and `spring.cloud.gcp.sql.jdbc.enabled`.

Reply:
So for this PR, the check would be that both `spring.cloud.gcp.sql.enabled` and `spring.cloud.gcp.sql.r2dbc.enabled` are true. That sounds good.

Reply:
Yep. Also, consider introducing `spring.cloud.gcp.sql.jdbc.enabled`.
Eventually, we might want to deprecate `spring.cloud.gcp.sql.enabled`.

Reply:
This is something that I did not think of before. Since it is possible for an app to expect a blocking connection while having the r2dbc driver on the classpath, introducing `spring.cloud.gcp.sql.r2dbc.enabled` sounds like a good idea.  I'll add `spring.cloud.gcp.sql.jdbc.enabled` in a separate PR since it might break existing behavior for current cloud sql customers.

Reply:
@meltsufin @elefeint Added `spring.cloud.gcp.sql.r2dbc.enabled` which will be `true` by default.